### PR TITLE
Classifier v0.8

### PR DESCRIPTION
**Pull request by snim2**
New classifier code, needs squashing.

### Test case

**Master branch**

  * [master_branch_bencher5_table.pdf](https://github.com/softdevteam/warmup_experiment/files/868349/master_branch_bencher5_table.pdf)
  * [master_branch_bencher6_table.pdf](https://github.com/softdevteam/warmup_experiment/files/868352/master_branch_bencher6_table.pdf)
  * [master_branch_bencher7_table.pdf](https://github.com/softdevteam/warmup_experiment/files/868353/master_branch_bencher7_table.pdf)

**New classifier**

  * [new_classifier_bencher5_table.pdf](https://github.com/softdevteam/warmup_experiment/files/868354/new_classifier_bencher5_table.pdf)
  * [new_classifier_bencher6_table.pdf](https://github.com/softdevteam/warmup_experiment/files/868355/new_classifier_bencher6_table.pdf)
  * [new_classifier_bencher7_table.pdf](https://github.com/softdevteam/warmup_experiment/files/868356/new_classifier_bencher7_table.pdf)

### CODE REVIEW

File: warmup/summary_statistics.py [hunk 1]
```diff
@@ -29,7 +29,8 @@ def collect_summary_statistics(data_dictionaries, half_bound, delta, steady_stat
     keys = sorted(data_dictionaries[machine]['wallclock_times'].keys())
     for key in sorted(keys):
         wallclock_times = data_dictionaries[machine]['wallclock_times'][key]
-        if len(wallclock_times) == 0:            print ('WARNING: Skipping: %s from %s (no executions)' %
+        if len(wallclock_times) == 0:
+            print ('WARNING: Skipping: %s from %s (no executions)' %
```
Comment (ltratt) [inline warmup/summary_statistics.py]: Space after `print`.

File: bin/mark_changepoints_in_json [hunk 2]
```diff
@@ -99,27 +102,29 @@ class Segments(object):
...
+        last_segment = self.segments[-1]
+        lower_bound = min(last_segment.mean - last_segment.variance,
+                          last_segment.mean - self.delta)
+        upper_bound = max(last_segment.mean + last_segment.variance,
+                          last_segment.mean + self.delta)
         classification = 'flat'
         for index in xrange(len(self.segments) - 2, -1, -1):
             current_segment = self.segments[index]
-            if (current_segment.mean >= lower_bound and current_segment.mean <= upper_bound):
+            if (current_segment.mean + current_segment.variance >= lower_bound and
```
Comment (ltratt) [inline bin/mark_changepoints_in_json:113]: This should, I think, be:
```python
if current_segment.mean >= lower_bound and current_segment.mean <= upper_bound:
```

File: warmup/summary_statistics.py [hunk 3]
```diff
@@ -59,11 +60,14 @@ def collect_summary_statistics(data_dictionaries, half_bound, delta, steady_stat
...
-                lower_bound = min(last_segment_mean * (1.0 - half_bound), last_segment_mean - delta)
-                upper_bound = max(last_segment_mean * (1.0 + half_bound), last_segment_mean + delta)
+                last_segment_var = data_dictionaries[machine]['changepoint_vars'][key][p_exec][-1]
+                lower_bound = min(last_segment_mean - last_segment_var, last_segment_mean - delta)
+                upper_bound = max(last_segment_mean + last_segment_var, last_segment_mean + delta)
                 for index in xrange(len(data_dictionaries[machine]['changepoint_means'][key][p_exec]) - 2, -1, -1):
                     current_segment_mean = data_dictionaries[machine]['changepoint_means'][key][p_exec][index]
-                    if current_segment_mean >= lower_bound and current_segment_mean <= upper_bound:
+                    current_segment_var = data_dictionaries[machine]['changepoint_vars'][key][p_exec][index]
+                    if (current_segment_mean + current_segment_var >= lower_bound and
```
Comment (ltratt) [inline warmup/summary_statistics.py:69]: Same comment as above.

File: bin/mark_changepoints_in_json (hunk 2 above)
Comment (ltratt) [reply bin/mark_changepoints_in_json:113]: Ignore me. I was wrong!

**Comment by ltratt** [comment]
Ignore my two comments. I think this is ready for squashing.

**Comment by snim2** [comment]
Splat.

**Review by ltratt — APPROVED** [review]

**Pull request merged**